[Pan Jing Bin] iP

IceBear789 opened the pull request.

Commit:
Add Gradle support

Commit:
build.gradle: Update version to 8.29

IceBear789 committed:
Add greet, echo and exit

IceBear789 committed:
Add add and list

IceBear789 committed:
Add mark as done

IceBear789 committed:
Add ToDo

IceBear789 committed:
Add Event

IceBear789 committed:
Add Deadline

Add all three level 4 functionality

IceBear789 committed:
Add automated text UI testing

IceBear789 committed:
Add the ability to handle errors

IceBear789 committed:
Add Delete functionality

IceBear789 committed:
Add Save functionality

Duke is unable to save the task list on the local computer itself.

The software might crash and cause data to be lost. Thus a backup
is needed.

Save the task list on the local computer serves as a backup in case
data is lost.

Saving the task list on the local computer is convenient and
also reliable.

IceBear789 committed:
Add ability to process date and times

Duke only sees date and time as just a string.

If duke can understand date and times properly, more advanced features
can be implemented.

Implement date in Duke as a java.time.LocalDate object.

Choose java.time.localDate as it has all the necessary features.

IceBear789 committed:
Refactor: Fixed some typos in error messages

IceBear789 committed:
Merge branch 'branch-Level-7'

IceBear789 committed:
Merge branch 'branch-Level-8'

Resolve merge conflicts regarding java.util.LocalDate

IceBear789 committed:
Add Parser class

Duke's ability to parse commands is very basic.

More advanced parsing features allow duke to understand commands better.

Implement a new Parser class to improve Duke's parsing ability.

Creating a separate Parser class adheres to OOP principles and makes
code more readable.

IceBear789 committed:
Add Storage and TaskList classes

Duke's task processing and file processing features are very basic.

We extend Duke's functionality by adding more task processing and file
processing features.

Implement 2 new classes Storage and TaskList.

Creating separate classes for each function adheres to OOP principles
and makes code more readable.

IceBear789 committed:
Add Ui class

Duke's Ui features are very basic.

We extend Duke's functionality by adding more Ui features.

Implement a new Ui class to handle Ui features.

Creating separate classes for each function adheres to OOP principles
and makes code more readable.

IceBear789 committed:
refactor: Organise classes into one main package

Duke's classes are not organised into packages.

Difficult to manage a large number of classes without packages.

Organise the classes into packages.

We can put related classes in the same package, making managing classes
easier.

IceBear789 committed:
Add JUnit tests for Parser Class

Duke currently does not have tests for Parser Class.

There is a high chance a bug may occur in the Parser class in the
future, when new code is added.

Implement JUnit tests to ensure that the Parser class still functions
correctly after future changes.

JUnit tests were chosen due to their convenience.

IceBear789 committed:
Add more JUnit tests for Parser class

IceBear789 committed:
Fix minor coding errors

IceBear789 committed:
Add JavaDoc comments

The code for Dukebot have no JavaDoc comments.

This makes certain parts of the code difficult to understand.

Add JavaDoc comments for better readibility.

IceBear789 committed:
Improve coding standard

Some parts of the code did not follow the standard coding style.

This makes certain parts of the code difficult to understand.

Standardise coding for better readibility.

IceBear789 committed:
Add Find

Duke does not have the ability to search task by keyword.

When there are a large number of tasks, finding a specific task can be
very tedious.

Implement find method to quickly find specific tasks by keyword.

Users can thus easily find specific tasks, even if there are a large
number of tasks.

IceBear789 committed:
Fix minor bugs

IceBear789 committed:
Merge branch 'branch-A-JavaDoc'

IceBear789 committed:
Merge branch 'branch-A-CodingStandard'

IceBear789 committed:
Merge branch 'branch-Level-9'

Keep changes in both branch-Level-9 and master branch.

IceBear789 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

IceBear789 committed:
Add Gradle

Duke project builds are not automated.

As the codebase gets larger, building Duke manually every single time
gets too time-consuming.

Automate build process of Duke using Gradle.

This speeds up the build process via automation.

IceBear789 committed:
Add JavaDocs

IceBear789 committed:
Merge branch 'branch-A-Gradle'

IceBear789 committed:
Refactor: Rewrote code that violates Checkstyle

IceBear789 committed:
Add JavaDocs

IceBear789 committed:
Merge branch 'branch-A-Checkstyle'

IceBear789 committed:
Add GUI for Duke

Duke does not have currently have a GUI.

Duke feels bland and boring.

Add a Graphical User Interface to Duke.

Makes Duke more graphically pleasing.

IceBear789 committed:
Refactor: Implement GUI using FXML

The GUI for Duke is not implemented using FXML.

Creating a GUI is long and tedious as it requires many lines of code.

Implement GUI using FXML.

Makes implementing a GUI faster and easier as FXML is more concise.

IceBear789 committed:
Add JavaDocs

IceBear789 committed:
Merge branch 'branch-Level-10'

NightRaven49 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,56 @@
+     * Overloaded constructor to create a deadline task object. It accepts one extra argument
+     * to determine if the task is already done.
+     *
+     * @param taskName The name of the deadline
+     * @param isDone Whether the deadline is already done
+     * @param deadline The date in which the deadline is due
```
Should these have punctuations?
Possible to review the other Javadocs in the rest of your code too.

xiinweii98 replied:
Strictly speaking, I remember each sentence should end properly with a full stop.

NightRaven49 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,139 @@
+     *
+     * @param taskList the corresponding task list in which the text file is based on
+     */
+    public static void update(ArrayList<Task> taskList) {
```
Should taskList be tasks instead?

NightRaven49 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,139 @@
+     * @throws IOException whenever an issue with reading the file arises
+     */
+    public static void convert(File file, ArrayList<Task> taskList) throws IOException {
+        Scanner scan = new Scanner(file);
```
Perhaps you can consider naming the Scanner object as a noun rather than a verb.

NightRaven49 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,163 @@
+ * tasks in the list.
+ */
+class TaskList {
+    private ArrayList<Task> taskList;
```
Do consider using plural forms for collections.

NightRaven49 reviewed (commented):
Overall, code is generally free of errors, though there is some inconsistency in naming and Javadocs. Was easy to read and understand what was the flow in logic, and was particularly impressed with the level of abstraction achieved with the parser. Good job!

xiinweii98 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,71 @@
+                break;
+            } else {
+                try {
+                    String[] processedText = iceBear.parser.processCommand(nextCommand);
```
I am pretty sure you can access your private fields instead of going through `iceBear.parser`. However, I am unsure which one you should go with.

xiinweii98 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,71 @@
+            }
+        }
+    }
+    public String[] processCommand(String input) throws DukeException {
+        return this.parser.processCommand(input);
+    }
+
+    public String process(String[] processedInput) {
+        return this.ui.processCommand(processedInput, this.taskList);
+    }
```
What is the difference between this two methods? I think it would be good to include some java doc comments or make the method names distinct.

xiinweii98 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,147 @@
+     */
+    public String[] processCommand (String input) throws DukeException {
+        String command = input.split(" ")[0];
+        if (command.equals("list")) {
+            return this.processList(input);
+        } else if (command.equals("done")) {
+            return this.processDone(input);
+        } else if (command.equals("delete")) {
+            return this.processDelete(input);
+        } else if (command.equals("todo")) {
+            return this.processToDo(input);
+        } else if (command.equals("deadline")) {
+            return this.processDeadLine(input);
+        } else if (command.equals("event")) {
+            return this.processEvent(input);
+        } else if (command.equals("find")) {
+            return this.processFind(input);
+        } else {
+            throw new DukeException("Invalid command");
+        }
```
A switch case should be apt here.

xiinweii98 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Creates a new text file so that the list of tasks can be saved on the hard disk.
+     */
+    public static void createFile() throws java.io.IOException {
```
I am not sure why you are using `java.io.IOException` instead of just `IOException`.

xiinweii98 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,139 @@
+        return result;
+    }
+
+    private static LocalDate dateConverter(String date) {
```
Is this necessary? What if users choose to input dates in another format? So, I would think it is better to make the user input a certain format like yyyy-mm-dd instead of catering to the different formats that a user can input.

xiinweii98 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,48 @@
+     * @param date the LocalDate object to print the date of.
+     * @return the string representation of the date of the object in the special format.
+     */
+    public static String printDate(LocalDate date) {
```
I think you can use the inbuilt methods to achieve your desired results. [https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/LocalDate.html#format(java.time.format.DateTimeFormatter)](url)

xiinweii98 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,163 @@
+ * The class that contains the task list. It has operations to add/delete
+ * tasks in the list.
+ */
+class TaskList {
```
It seems like your TaskList is doing the job of Ui returning String objects to be printed/shown to user.

xiinweii98 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,37 @@
+     */
+    protected String processCommand(String[] input, TaskList taskList) {
+        try {
+            if (input[0].equals("list")) {
+                return taskList.listTask();
+            } else if (input[0].equals("done")) {
+                return taskList.doneTask(Integer.valueOf(input[1]));
+            } else if (input[0].equals("delete")) {
+                return taskList.delete(Integer.valueOf(input[1]));
+            } else if (input[0].equals("todo")) {
+                return taskList.addToDo(input[1]);
+            } else if (input[0].equals("deadline")) {
+                return taskList.addDeadline(input[1], LocalDate.parse(input[2]));
+            } else if (input[0].equals("event")) {
+                return taskList.addEvent(input[1], LocalDate.parse(input[2]));
+            } else if (input[0].equals("find")) {
+                return taskList.find(input[1]);
+            }
```
A switch case would be apt here.

xiinweii98 reviewed (commented):
Overall nicely written code. Keep up the good work!

IceBear789 committed:
Refactor: Improve code quality and add more comments

IceBear789 committed:
Refactor: Improve code quality

IceBear789 committed:
Add: More duke images

IceBear789 committed:
Revert "Add: More duke images"

This reverts commit c87390b8fb7f823baf73787b9476b8a947696d01.

IceBear789 committed:
Merge pull request #1 from IceBear789/branch-A-CodeQuality

Improve code quality

IceBear789 committed:
Merge branch 'master' of https://github.com/IceBear789/ip

* 'master' of https://github.com/IceBear789/ip:
  Refactor: Improve code quality
  Refactor: Improve code quality and add more comments

IceBear789 committed:
Add assertions

There are no assertions in the codebase for dukebot.

As the codebase gets larger and larger, having no assertions may make
the program more difficult to debug.

Add assertions to the codebase of dukebot.

This helps us narrow down on the things to look out for when debugging

IceBear789 committed:
Add continuous integration

IceBear789 committed:
Refactor: Improve JavaDoc comments

IceBear789 committed:
Merge pull request #2 from IceBear789/branch-A-Assertions

Merge assertions

IceBear789 committed:
Merge branch 'master' of https://github.com/IceBear789/ip

* 'master' of https://github.com/IceBear789/ip:
  Add assertions

IceBear789 committed:
Merge pull request #3 from IceBear789/branch-A-CI

Add continuous integration

IceBear789 committed:
Add C-Sort

Duke does not currently have the function of sorting tasks.

This makes it difficult to keep track of certain tasks, especially when
the number of tasks get very large.

Implement the ability to sort tasks based on certain criteria.

Enable users to keep track of more important tasks easily.

IceBear789 committed:
Refactor : Improve Code Quality

IceBear789 committed:
Fix minor bugs

IceBear789 committed:
Add better GUI

IceBear789 committed:
Refactor: Improve code quality

IceBear789 committed:
Add user guide

IceBear789 committed:
Update user guide

IceBear789 committed:
Refactor: Improve code quality

IceBear789 committed:
Update user guide

IceBear789 committed:
Update docs

IceBear789 committed:
Update docs

IceBear789 committed:
Fix Junit tests

IceBear789 committed:
Refactor: Improve code quality

IceBear789 committed:
Fix typos in docs

IceBear789 committed:
Fix minor typos

IceBear789 committed:
Update text UI tests

IceBear789 committed:
Fix bugs
